import "astatus_code.proto";

package raft_demo;

option cc_generic_services = true;//FIX ME

message EchoRequest {
    optional string echo = 1;
}

message EchoResponse {
    optional string echo = 1;
    optional StatusCode status = 2;
}

message PutRequest{
    optional string key = 1;
    optional string value = 2;
}

message PutResponse {
    optional StatusCode status = 1;
}

message GetRequest {
    optional string key = 1;
}

message GetResponse {
    optional string value = 1;
    optional StatusCode status = 2;
}

message PutBatchRequest {
    optional int32 kv_nums = 1;
    optional bytes databuf = 2;
}

message PutBatchResponse {
    optional StatusCode status = 1;
}

message GetBatchRequest {
    optional string key_start = 1;
    optional string key_end = 2;
}

message GetBatchResponse {
    optional bytes databuf = 1;
    optional StatusCode status = 2;
}

message GetNodeDataRequest {
    optional uint32 key_hash_start = 1;
    optional uint32 key_hash_end = 2;
}

message GetNodeDataResponse {
    optional string databuf = 1;
    optional StatusCode status = 2;
}

service ServerService {
    rpc Echo(EchoRequest) returns(EchoResponse);
    rpc Put(PutRequest) returns(PutResponse);
    rpc Get(GetRequest) returns(GetResponse);
    rpc PutBatch(PutBatchRequest) returns(PutBatchResponse);
    rpc GetBatch(GetBatchRequest) returns(GetBatchResponse);
    rpc GetNodeData(GetNodeDataRequest) returns(GetNodeDataResponse);
}
